### PR TITLE
nixos/tests/k3s: add etcd backend

### PR DESCRIPTION
**Pull request by superherointj**
nixos/tests/k3s: add etcd backend
nixos/tests/etcd: add k3s

I thought of this test as desirable to be executed on updates/changes to etcd and k3s.
Because it can validate each other. I tried to keep the test very minimalist. But let me know what you guys think.

> nix build .#k3s.passthru.tests.etcd
> nix build .#etcd.passthru.tests.k3s

CC maintainers: @bryanasdev000
* etcd: @offline @endocrimes
* k3s: @euank @mic92 @yajo 

Thanks in advance for your wiser insights.

## This PR is ready to be merged

### CODE REVIEW

File: nixos/tests/k3s/etcd.nix
Comment (superherointj) [inline nixos/tests/k3s/etcd.nix:43]:
```suggestion
          "--pause-image" "test.local/pause:local"
```
@euank Thoughts?

File: nixos/tests/k3s/etcd.nix
Comment (bryanasdev000) [inline nixos/tests/k3s/etcd.nix:12]: Listening to 0.0.0.0 won't be better?

File: nixos/tests/k3s/etcd.nix
Comment (bryanasdev000) [inline nixos/tests/k3s/etcd.nix:74]: K3S auth (kubeconfig) is configured before these actions?

**Review by bryanasdev000 — APPROVED** [review]
Overall, LGTM, just a few nitpicks/ideas.

Backup and restore can be tested? (I know that K3S has its own mechanism, but it may help to use default etcdctl CLI like https://etcd.io/docs/v3.5/op-guide/recovery/)

**Review by yajo — APPROVED** [review]
Very good!

File: nixos/tests/k3s/etcd.nix
Comment (superherointj) [reply nixos/tests/k3s/etcd.nix:74]: No need.

File: nixos/tests/k3s/etcd.nix
Comment (superherointj) [reply nixos/tests/k3s/etcd.nix:12]: Better I don't know, but that would work too. I don't think there is a right answer here.

File: nixos/tests/k3s/etcd.nix
Comment (superherointj) [reply nixos/tests/k3s/etcd.nix:100]: @offline @endocrimes Would adding etcd maintainers to this test be acceptable?

File: nixos/tests/k3s/etcd.nix
Comment (euank) [reply nixos/tests/k3s/etcd.nix:43]: Mind explaining more what you mean? I don't see any difference in the diff you posted :sweat_smile:

File: nixos/tests/k3s/etcd.nix
Comment (euank) [inline nixos/tests/k3s/etcd.nix:87]: This feels a little odd to test.

I think the etcd structure k8s uses isn't guaranteed and can be broken, and for example has been in the past with the etcd 2-> 3 migration.

Because of that, it feels a little fragile to read this structure out of etcd.

I do get the appeal of verifying it's using the etcd backend, and I don't think they're going to change the structure anytime soon (it feels like it might be ossified at this point), but it still feels like a wrong thing to test.

I'm not going to block on this, just observe that I think k8s is within its right to break this test.

File: nixos/tests/k3s/etcd.nix
Comment (euank) [inline nixos/tests/k3s/etcd.nix:84]: We have a mix of `test` and `[[ ... ]]` for these sorts of assertions; I feel like we should stick with one for consistency, and I think `[[` is easier to read.

**Review by euank — COMMENTED** [review]
Overall looks good!
It's handy to have coverage for another way of running k3s, and more coverage for etcd too, thanks!

I left a couple comments, but I wouldn't consider any of them to be blockers for merging this.

**Comment by superherointj** [comment]
> "--pause-image" "test.local/pause:local"

Is this desirable? What practical purpose it serves?

**Comment by euank** [comment]
Ahh, gotcha.

The pause image thing is there in the other tests to use this image:

https://github.com/superherointj/nixpkgs/blob/45069973995893aecb9472233b726b2dba801ca4/nixos/tests/k3s/single-node.nix#L70-L72

At the time the test was written, that was necessary to run a pod at all since otherwise the test would try, and fail, to download the pause image from the network.

It looks like the pause image stuff has since changed significantly upstream (i.e. [this](https://github.com/kubernetes/kubernetes/issues/98765) and related threads), and as such it's probably not necessary anymore.

It's definitely not necessary for this test if we only create/delete a secret.

File: nixos/tests/k3s/etcd.nix
Comment (superherointj) [reply nixos/tests/k3s/etcd.nix:87]: As an integration test, both sides need to be sanity checked.
Whenever the resource path changes, we have to update it. But then, we can trust that k3s is actually using etcd (and not sqlite). Otherwise, I don't know how to test if etcd is actually being used.
I'm open to alternative implementations.

Comment (superherointj) [reply nixos/tests/k3s/etcd.nix:86]: > This feels a little odd to test.
> k8s is within its right to break this test.

A thought to consider: A test automates our debug process. Whenever something (`k8s`) breaks our debug process, we should update our debug process. So that breakage is welcomed because our debug process is stale.

I see a beauty in this. :-)

Comment (euank) [reply nixos/tests/k3s/etcd.nix:86]: The thing we really want confidence for here is that the nixos module / k3s configured to use etcd are actually using etcd, and the exposed user functionality works.

If this was an integration test for k3s, testing that etcd's contents are what's expected would make sense since that's part of k3s's integration with etcd.
In NixOS's case, the surface area we care about is just that the k3s config options make k3s actually use etcd, and the functionality of k3s works.

This does test that k3s is actually using etcd well, so in that sense it's a good test for sure! And in practice, I don't know of another way to test that, so this might be our best bet.

I didn't realize it tested something you actually used yourself as well, which is another point in factor of the test being useful.

Anyway, taking it all together, I'm fine merging this test. In practice I expect it to be stable, and if it turns out k8s does change the etcd structure more often than, say, once every 2 years, we can reconsider the test then.

Thanks for writing it!

Comment (superherointj) [reply nixos/tests/k3s/etcd.nix:86]: I get that you are **not** in opposition to this. But just wanted to clarify a nuance. Just so we are at same page. I think we should eventually agree on this. Because it is a logical conclusion. And likely I am failing to communicate it properly.

> If this **was** an integration test for k3s

I think this is still an integration test. The nuance is:
Upstream k3s tests their k3s with their etcd.
Here is being tested nixpkgs's k3s with nixpkgs's etcd (using NixOS configuration module).

> if ... k8s ... change the etcd structure ... we can **reconsider the test** then.

If this test ever gets removed, it's not possible to guarantee etcd side is working as expected.
K3s could be using it's embedded sqlite and we would never know.
On breakage, I think, the test should be updated to reflect the new debug process.

Comment (euank) [reply nixos/tests/k3s/etcd.nix:86]: Yup, understood. I think we agree overall.

Being sure that k3s is using etcd is a critical part of the test, and this is the mechanism we've picked for that.

Thanks for talking through it more!

**Review by yajo — APPROVED** [review]
Awesome

**Comment by superherointj** [comment]
Maybe wegank?

**Review by bryanasdev000 — APPROVED** [review]
Diff LGTM.

**Pull request merged**
